src/io/ResultWriter.cpp

```diff
@@ -1,89 +1,116 @@
-			<< std::setw(_writePrecision+15) << std::scientific << std::setprecision(_writePrecision) << domain->getGlobalBetaRot()
-			<< std::setw(_writePrecision+15) << std::scientific << std::setprecision(_writePrecision) << cv
-			<< std::setw(_writePrecision+15) << std::scientific << std::setprecision(_writePrecision) << globalNumMolecules
-			<< endl;
+	double uKinLocal = 0.0F;
+	double uKinGlobal = 0.0F;
+	double uKinTransLocal = 0.0F;
+	double uKinTransGlobal = 0.0F;
+	double uKinRotLocal = 0.0F;
+	double uKinRotGlobal = 0.0F;
+	#if defined(_OPENMP)
+	#pragma omp parallel reduction(+:uKinLocal,uKinTransLocal,uKinRotLocal)
+	#endif
+	{
+		for (auto moleculeIter = particleContainer->iterator(ParticleIterator::ONLY_INNER_AND_BOUNDARY);
+			 moleculeIter.isValid(); ++moleculeIter) {
+			uKinLocal += moleculeIter->U_kin();
+			uKinTransLocal += moleculeIter->U_trans();
+			uKinRotLocal += moleculeIter->U_rot();
+		}
 	}
-}
+#ifdef ENABLE_MPI
+    MPI_Allreduce(&uKinLocal, &uKinGlobal, 1, MPI_DOUBLE, MPI_SUM, domainDecomp->getCommunicator());
+	MPI_Allreduce(&uKinTransLocal, &uKinTransGlobal, 1, MPI_DOUBLE, MPI_SUM, domainDecomp->getCommunicator());
+	MPI_Allreduce(&uKinRotLocal, &uKinRotGlobal, 1, MPI_DOUBLE, MPI_SUM, domainDecomp->getCommunicator());
+#else
+    uKinGlobal = uKinLocal;
+	uKinTransGlobal = uKinTransLocal;
+	uKinRotGlobal = uKinRotLocal;
+#endif
+	_uKin_acc += uKinGlobal;
+	_uKinTrans_acc += uKinTransGlobal;
+	_uKinRot_acc += uKinRotGlobal;
 
-void ResultWriter::finish(ParticleContainer * /*particleContainer*/,
-						  DomainDecompBase *domainDecomp, Domain * /*domain*/){
+	_numSamples++;
 
-	if(domainDecomp->getRank() == 0) {
-		const auto now = std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
-		tm unused{};
-		const auto nowStr = std::put_time(localtime_r(&now, &unused), "%c");
-		_resultStream << "# ls1 mardyn simulation finished at " << nowStr << endl;
-		_resultStream.close();
+	if ((simstep % _writeFrequency == 0) and (simstep > 0UL)) {
+		// Only main rank writes data to file
+		if (domainDecomp->getRank() == 0) {
+			const string resultfile(_outputPrefix+".res");
+			std::ofstream resultStream;
+			resultStream.open(resultfile.c_str(), std::ios::app);
+			auto printOutput = [&](auto value) {
+				resultStream << std::setw(_writeWidth) << std::scientific << std::setprecision(_writePrecision) << value;
+			};
+			resultStream << std::setw(10) << simstep;
+				printOutput(_simulation.getSimulationTime());
+				printOutput(_uPot_acc/_numSamples);
+				printOutput(_uKin_acc/_numSamples);
+				printOutput(_uKinTrans_acc/_numSamples);
+				printOutput(_uKinRot_acc/_numSamples);
+				printOutput(_p_acc/_numSamples);
+				printOutput(domain->cv());
+				printOutput(globalNumMolecules);
+				resultStream << std::endl;
+			resultStream.close();
+		}
+
+		// Reset values
+		_numSamples = 0UL;
+		_uPot_acc = 0.0F;
+		_uKin_acc = 0.0F;
+		_uKinTrans_acc = 0.0F;
+		_uKinRot_acc = 0.0F;
+		_p_acc = 0.0F;
```

Comment:
The existing implementation was not what was [documented](https://github.com/ls1mardyn/ls1-mardyn/blob/b6f274e625fe7a8c07861f09c2b6c9c5e0243f2e/src/io/ResultWriter.h#LL13C45-L13C45). For me, it seems like the accumulator returns not only the value averaged over the last `_writeFrequency` steps but for the whole simulation. Otherwise it would had to reset the data after getting the average which it does not.
What do you mean by "higher write frequency than the sample count"? The sample count just counts the number of sampling steps. It is often the writing frequency but not at the very beginning (simstep=0 is sampled but it would lead to wrong averages if at e.g. simstep=10=writefrequency) the accumulated value is divided by 10 and not 11.)
I not that it is not perfect to have the loop and MPI call in here. But at the moment, the domain/ensemble does not provide a working method to get the global kinetic energy.

Reply:
@cniethammer So what do you suggest regarding my 3 comments?